sys/event/thread.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2019 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2018 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_event
+ * @{
+ *
+ * @file
+ * @brief       Event Loop Thread implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "thread.h"
+#include "event.h"
+#include "event/thread.h"
+
+static void *_handler(void *event_queue)
+{
+    event_queue_claim(event_queue);
+    event_loop(event_queue);
+
+    /* should be never reached */
+    return NULL;
+}
+
+void event_thread_init(event_queue_t *queue, char *stack, size_t stack_size,
+                       unsigned priority)
+{
+    /* For the auto_init use case, this will be called before main gets
+     * started.  main might already use the queues, so they need to be
+     * initialized at that point.
+     *
+     * They will be claimed within the handler thread.
+     */
+    event_queue_init_detached(queue);
+
+    thread_create(stack, stack_size, priority, 0, _handler, queue, "event");
+}
+
+#ifndef EVENT_THREAD_HIGHEST_STACKSIZE
+#define EVENT_THREAD_HIGHEST_STACKSIZE ISR_STACKSIZE
+#endif
+#ifndef EVENT_THREAD_HIGHEST_PRIO
+#define EVENT_THREAD_HIGHEST_PRIO   (0)
+#endif
+
+#ifndef EVENT_THREAD_LOWEST_STACKSIZE
+#define EVENT_THREAD_LOWEST_STACKSIZE ISR_STACKSIZE
+#endif
+#ifndef EVENT_THREAD_LOWEST_PRIO
+#define EVENT_THREAD_LOWEST_PRIO   (THREAD_PRIORITY_IDLE - 1)
```

Comment:
We would have at most exactly two priorities and two threads, right? Shouldn't the lowest thread priority be `(THREAD_PRIORITY_MAIN - 1)`? Otherwise drivers would have to use again the high priority thread to ensure that their events are at least handled before the main thread is executed.

Reply:
> Otherwise drivers would have to use again the high priority thread to ensure that their events are at least handled before the main thread is executed.

Maybe they want to, as in, "do this when there is time". That's why I've made it just one higher than idle.

How about I add a third priority ("medium"), which is just one higher than main? Or does the current very-low one make sense at all?

I mean, this is just the auto configuration, cpus / boards / applications can freely launch more handler threads...

Reply:
> How about I add a third priority ("medium"), which is just one higher than main? Or does the current very-low one make sense at all?

What priority should a thread have that is used by a driver to handle an interrupt from the hardware? IMHO, its priority should be at least higher than the priority of the main thread. A medium priority thread seems to be more important than a very low priority thread.

> I mean, this is just the auto configuration, cpus / boards / applications can freely launch more handler threads...

Yes of course, but drivers can only use threads that are known from the auto configuration. Otherwise, they would need an additional thread parameter for the initializtion. This in turn would prevent the auto initialization.

Reply:
> What priority should a thread have that is used by a driver to handle an interrupt from the hardware?

I guess that depends... It is easy to add more priorities, so I propose I add one more middle priority (one higher than main) and see how far we go with that.

Reply:
:+1: The three priorities high (0), medium (main-1) and low (idle-1) probably meet the requirements of most use cases.

Reply:
added medium priority